### PR TITLE
[BEAM-8953] Extend ParquetIO read builders for AvroParquetReader

### PR DESCRIPTION
**Pull request by RyanBerti**
**Please** add a meaningful description for your change here

Extends read / readFiles builders to support an extra argument, allowing users to configure the data model that the ParquetAvroReader's wrapped AvroReadSupport class uses. We're running these changes locally, and they allow us to use avro4s to convert from GenericRecords to Scala case classes without having to extend IndexedRecords/SpecificData (we're using Scio but utilizing Beam's ParquetIO rather than Scio's Parquet support). I'd like to get some feedback from the Parquet owners on the approach; I'm not sure if there's a better way to drive ParquetAvroReader configuration from Beam applications. This is my first PR against the Beam project, so let me know if there's anything I missed.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by RyanBerti** [comment]
looks like @lgajowy @jbonofre @aromanenko-dev are listed as the owners of the parquet extension; would be happy to get feedback based on the PR.

**Comment by RyanBerti** [comment]
Link to JIRA https://issues.apache.org/jira/browse/BEAM-8953

**Comment by aromanenko-dev** [comment]
Thank you for contribution! 
I'll take a look on this. Just a question - could you add unit tests for that?

**Comment by RyanBerti** [comment]
@aromanenko-dev ya I'll definitely add a unit test; just wanted to get some initial feedback on whether you guys think there's a better path here or not.

File: sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java [hunk 1]
```diff
@@ -126,6 +127,15 @@ public static Read read(Schema schema) {
     return new AutoValue_ParquetIO_Read.Builder().setSchema(schema).build();
   }
 
+  /**
+   * Like {@link #read(Schema)}, with an additional {@link org.apache.avro.generic.GenericData}
+   * argument, allowing users to signal to the underlying AvroParquetReader the type of model to
+   * associate with the wrapped {@link org.apache.parquet.avro.AvroReadSupport} instance.
+   */
+  public static Read read(Schema schema, GenericData model) {
```
[the remark this answers is not in the document]
Comment (aromanenko-dev) [reply sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java]: Agree with @lgajowy - it should be just `withXXX` method since this option is not required for all pipelines.

[the remark this answers is not in the document]
Comment (RyanBerti) [reply sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java]: @lgajowy  @aromanenko-dev PR updated to support the proposed chaining method for withDataModel; added unit tests showing how avro fails if the schema is built from a class.

File: sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java [hunk 2]
```diff
@@ -144,6 +166,9 @@ public static ReadFiles readFiles(Schema schema) {
     @Nullable
     abstract Schema getSchema();
 
+    @Nullable
+    abstract GenericData getAvroDataModel();
```
[the remark this answers is not in the document]
Comment (RyanBerti) [reply sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java]: @aromanenko-dev I don't see that Schema is Serlializable; is it just the components of the DoFn that need to be Serializable? If so, Class is Serializable. Is there anyway I can validate whether these additions will cause serialization problems or not via a unit test?

[the remark this answers is not in the document]
Comment (aromanenko-dev) [reply sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java]: Serialisation/Deserialisation should be tested by DirectRunner when you run your test with TestPipeline. My concern is that `GenericData` is a part of AutoValue class which could be copied over network to workers, so it should serialised/deserialised in this case.

**Comment by RyanBerti** [comment]
@aromanenko-dev @lgajowy Just wanted to ping you guys to see if you'd have a chance to take a look at the updated PR in the next few weeks?

**Comment by RyanBerti** [comment]
@aromanenko-dev would you mind taking another pass at this? Thanks!

File: sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (RyanBerti) [reply sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java]: @aromanenko-dev sorry I didn't see this comment; as I pointed out, GenericData would join Schema as instance variables of the AutoValue annotated ReadFiles class (neither of which are serializable). From the docs, only the ParDo itself must be serializable (https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ParDo.java#L282). When the ParDo is created anonymously, the encompassing PTransform needs to be serializable, but that's not the case with ParquetIO. I've run the changes both locally via the DirectRunner and on Cloud Dataflow without issue; I was asking if there were a way to unit test the serialization requirements, but that doesn't seem to be an option. Let me know if building out serialization for all of the components of the PTransform is required, or if we can move ahead without serialization of the Schema and GenericData instance vars.

[the remark this answers is not in the document]
Comment (aromanenko-dev) [reply sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java]: I agree that a definition of serialisation of PTransform objects is not very clear. After a discussion on [dev@](https://lists.apache.org/thread.html/ree6bb28e0df524efffb220db526edc83ea0dbecdd537e3e5d9cb058c%40%3Cdev.beam.apache.org%3E) I may confirm that it's not required to be serialisable for all members of PTransform class if they are not reachable by DoFn or any other user-defines objects that will be executed on workers. Seems like, in your case you call `getAvroDataModel()` only on driver, so it should be fine to have `GenericData` is not `Serializable`.

File: sdks/java/io/parquet/src/test/java/org/apache/beam/sdk/io/parquet/ParquetIOTest.java [hunk 3]
```diff
@@ -126,4 +128,57 @@ public void testReadDisplayData() {
...
+
+  public static class TestRecord {
+    String name;
+
+    public TestRecord(String name) {
+      this.name = name;
+    }
+  }
+
+  @Test(expected = org.apache.beam.sdk.Pipeline.PipelineExecutionException.class)
```
Comment (aromanenko-dev) [inline sdks/java/io/parquet/src/test/java/org/apache/beam/sdk/io/parquet/ParquetIOTest.java]: Does it fail only in case if we obtain schema using Avro `ReflectData`?

File: sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java [hunk 4]
```diff
@@ -165,6 +171,10 @@ public Read from(String filepattern) {
       return from(ValueProvider.StaticValueProvider.of(filepattern));
     }
 
+    public Read withAvroDataModel(GenericData model) {
```
Comment (aromanenko-dev) [inline sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java]: Could you add Javadoc for this public method and maybe an example into main class?

File: sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java [hunk 5]
```diff
@@ -192,21 +202,40 @@ public void populateDisplayData(DisplayData.Builder builder) {
...
     @AutoValue.Builder
     abstract static class Builder {
       abstract Builder setSchema(Schema schema);
 
+      abstract Builder setAvroDataModel(GenericData model);
+
       abstract ReadFiles build();
     }
 
+    public ReadFiles withAvroDataModel(GenericData model) {
```
Comment (aromanenko-dev) [inline sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java]: Could you add Javadoc for this public method and maybe an example into main class?

**Comment by aromanenko-dev** [comment]
@RyanBerti It's almost LGTM for me, I just left several comments, PTAL

File: sdks/java/io/parquet/src/test/java/org/apache/beam/sdk/io/parquet/ParquetIOTest.java (hunk 3 above)
Comment (RyanBerti) [reply sdks/java/io/parquet/src/test/java/org/apache/beam/sdk/io/parquet/ParquetIOTest.java]: For schemas that include a namespace and name that map back to a valid class in the classpath, the AvroRecordConverter.start() method will attempt to create an instance of the given class when the SpecificData model is in use (this doesn't happen when the name/namespace aren't set). So yes - when the schema is obtained with ReflectData, the name and namespace would be 'TestRecord' and 'beam.sdks.java.io.parquet.test', and since this class doesn't implement a no-arg constructor and the other methods expected by SpecificData the runner ends up throwing a NoSuchMethodException.

File: sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java (hunk 4 above)
Comment (RyanBerti) [reply sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java]: Will do.

File: sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java (hunk 5 above)
Comment (RyanBerti) [reply sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java]: Will do.

**Comment by RyanBerti** [comment]
@aromanenko-dev added javadocs - let me know if it's too wordy / not wordy enough?

**Comment by aromanenko-dev** [comment]
retest this please

**Comment by aromanenko-dev** [comment]
@RyanBerti Thanks, LGTM. I just want to run tests again and we can merge if it's green

**Comment by aromanenko-dev** [comment]
@RyanBerti There is an issue with `spotless` plugin (see [jenkins log](https://builds.apache.org/job/beam_PreCommit_Spotless_Commit/7153/console)). Could you fix it by running `./gradlew :sdks:java:io:parquet:spotlessApply` and update your PR?
Also, it's recommended to run `check` task on changed modules, like `./gradlew :sdks:java:io:parquet:check`, before commit.

**Comment by RyanBerti** [comment]
@aromanenko-dev sorry about that! pushed another commit to fix

**Comment by RyanBerti** [comment]
retest this please

**Comment by aromanenko-dev** [comment]
retest this please

**Comment by aromanenko-dev** [comment]
retest this please

**Comment by aromanenko-dev** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by aromanenko-dev** [comment]
Run Java_Examples_Dataflow PreCommit
